github/commands/pull_request.py

```diff
@@ -121,12 +124,48 @@ def create_branch_for_pr(self, branch_name):
             self.pr["head"]["ref"]
         )
 
+        if not branch_name:
+            branch_name = self.pr["head"]["sha"]
+
         self.window.status_message("Creating local branch for PR...")
         self.git(
             "branch",
             branch_name,
             self.pr["head"]["sha"]
         )
+        sublime.set_timeout(lambda: self.set_upstream_for_branch(branch_name))
+
+    def set_upstream_for_branch(self, branch_name):
+        remotes = list(self.get_remotes().keys())
+        remote = self.pr["user"]["login"]
+        remote_branch = self.pr["head"]["ref"]
+
+        clone_url = self.pr["head"]["repo"]["clone_url"]
+        ssh_url = self.pr["head"]["repo"]["ssh_url"]
+
+        if remote not in remotes:
+            if not sublime.ok_cancel_dialog("Add remote '{}'?".format(remote)):
+                return
+
+            def on_select_url(index):
+                if index < 0:
+                    return
+                elif index == 0:
+                    url = clone_url
+                elif index == 1:
+                    url = ssh_url
+
+                self.git("remote", "add", remote, url)
+                self.git("fetch", remote, remote_branch)
+                self.git("branch", "-u", "{}/{}".format(remote, remote_branch), branch_name)
+
+            sublime.set_timeout(
+                lambda: self.window.show_quick_panel(
+                    [clone_url, ssh_url], on_select_url)
+            )
+        else:
+            self.git("fetch", remote, remote_branch)
```

Comment:
Here and above L159: we actually already fetched at the beginning of this action. 

```
        self.git(
            "fetch",
            self.pr["head"]["repo"]["clone_url"],
            self.pr["head"]["ref"]
        )
```

Why is that needed? Is there any difference? Not if the user selects `clone_url`, if the user selects `ssh` it just uses a different transport protocol, the ref is the same, and we already have it. No?

Reply:
it is needed for "Checkout as detached HEAD." where we do need to set upstream. Perhaps we should prompt user to choose what protocol to fetch.

Reply:
Now we always prompt user to choose a protocol when checkout as detached head.

Reply:
Why is the protocol a matter here? It never was before. As I said these days using https is just fine. Is there an issue open requesting and explaining the need? If this is a real issue we should deduce: if the user already used only non https for the remotes, we do it here as well. Otherwise we use https.

It confuses me to set an upstream if the user chooses detached head. That doesn't make any sense to me. We could ask to add the remote, but iirc the user can just checkout the same PR as a branch later if she needs to add commits. I mean, you only need the remote and upstream if you want to edit the PR, and that doesn't play well with a detached HEAD.

Reply:
Ah you already changed that and don't set the upstream in the detached case. So the only question is why the protocol matters now.

Reply:
In Linux, I think it is more common/easier to use ssh key rather than `credential.helper`. 

Also, the https protocol is particularly painful if two-factor authentication is used.

Reply:
I feel users should be allowed to choose, regardless of github recommendation. Also, it's important that we don't break any behaviour for existing users

Reply:
The user already made that choice. We can take whatever is set up before. Just look at the already configured remote urls and go with https if it's in there. I don't see why I should make this decision here over and over again. If you made the initial clone using https, just stick with it.

Reply:
If the remote exists, you won't be prompted to choose a URL. However, user will be asked if they are fetching the PR as detached head or the remote doesn't exist.

Reply:
The remote of the main repo determines which protocol I want to use and have set up. Since we're checking out a fork here there must be at least the main repo already configured.

Reply:
Perhaps the easier solution is to make a setting `pull_request_protocol` with default value `"ask"` and possible values "ssh" and "https".

Reply:
No that's the most complicated solution. Doesn't `self.get_integrated_remote_url()` already gives us the current base remote url, we query for this at the beginning of the process, and so we just look if `base_remote.starts_with("https:")`.

Reply:
Alright. It makes sense. I believe that most people don’t need to switch between protocol. I have to go sleeping now, will do the changes tomorrow.